### ISSUE TITLE
[cshake library] Choice of cSHAKE library in python

### ISSUE THREAD

**Issue by msfschaffner** [lowRISC/opentitan#6694]
We need a cSHAKE Python implementation in order to enable our scripting flows for ROM checking and life cycle token hashing. Unfortunately, both the built-in `hashlib` (https://pypi.org/project/pycryptodome/) and the `pycryptodome` package (https://pypi.org/project/pycryptodome/, part of our Python requirements) do not natively support cSHAKE. There is support for SHAKE and other SHA-3 modes in these libs, but as far as I understand we cannot reuse these modes as they transparently apply different bit padding strings (`0x1F` etc) than cSHAKE (`0x04`), and these strings cannot be overridden through the function interfaces:

https://github.com/python/cpython/blob/7148293d96843ca868961313b00361504ec0c242/Modules/_sha3/kcp/KeccakHash.h#L51-L73

https://github.com/Legrandin/pycryptodome/blob/5dace638b70ac35bb5d9b565f3e75f7869c9d851/lib/Crypto/Hash/SHAKE128.py#L57

So possible ways forward would be:
1) Search for another implementation that is not part of a standard package.
2) Get cSHAKE support added to a standard package.

As for 1), I found https://github.com/Hemoth/cSHAKE, which could be an option.

However, I am more leaning towards 2) since:

- the difference between SHAKE and cSHAKE is very small, and basically boils down to a slightly different hash object constructor (see comments on https://bugs.python.org/issue39539)
- the open source community will benefit from this being added to a mainstream package. also, there seems to be demand for it (see again https://bugs.python.org/issue39539)

I took a look at both the `hashlib` and `pycryptodome` packages, and it looks like the latter is a bit more accessible and straightforward to modify (the former interfaces to a C++ implementation underneath, and requires quite a bit of integration magic). Further, `hashlib` is part of cPython, which means that it may be more difficult to actually get this upstreamed in reasonable time. `pycrypotodome` is a standalone package that we can modify and host under lowrisc until the patch has landed upstream.

I thus went ahead and made a prototype implementation for `pycryptodome`, which appears to be working.
I need to clean this up a bit and will post a link to the code here shortly.

**Comment by msfschaffner** [comment]
See prototype implementation here: https://github.com/lowRISC/pycryptodome/pull/1

**Comment by moidx** [comment]
+1 for adding cSHAKE support to the standard package.

**Comment by cdgori** [comment]
Agreed, adding to standard package is good all around, and choosing `pycryptodome` over `hashlib` seems reasonable to me.

**Comment by msfschaffner** [comment]
With #6728 merged we now have support for cSHAKE functions via a lowRISC hosted development fork of `pycryptodome`. I'll leave this issue open until the PR upstream (https://github.com/Legrandin/pycryptodome/pull/531) has landed, and we've updated our Python requirements accordingly.

**Comment by e7p** [comment]
Hint until the Pull Request is merged into pycryptodome upstream: You can install the cSHAKE additions using this command:
`python3 -m pip install -e git+https://github.com/lowRISC/pycryptodome.git#egg=pycryptodome`

**Comment by msfschaffner** [comment]
Thanks for the hint, @e7p. Indeed, we currently host a development fork under lowrisc and plan to replace that once the cSHAKE additions have made it into a release:
https://github.com/lowRISC/opentitan/blob/93536cde3ccca091fac73a0304db54aea4ad32af/python-requirements.txt#L54-L57

**Comment by imphil** [comment]
The relevant changes are now merged upstream, a new release should appear any day now, allowing us to switch back to a released version of pycryptodome.
